### PR TITLE
add nkrokeyboard support

### PR DESCRIPTION
**Pull request by hlcmOxO**

### CODE REVIEW

**Review by hathach — COMMENTED** [review]
hi hi, sorry for late response. I was a bit busy with other works. I saw your issue #205 and #206 . Indeed there is a hipcup with esp32 where it use an older header, we can add an skip file to skip test for esp32 in ci (funhouse, magtag and metroesp32s2) like this example https://github.com/adafruit/Adafruit_TinyUSB_Arduino/tree/master/examples/MassStorage/msc_sd

The rest (samd/nrf) seems to be explicit type case which we can fix with the example. If you are still not sure how to do that, don't worry, just add back the ino file. I will take care of the ci (by pushing fix myself)

**Comment by hlcmOxO** [comment]
> hi hi, thank you for your PR and sorry for late response. I was a bit busy with other works. I saw your issue #205 and #206 . Indeed there is a hipcup with esp32 where it use an older header, we can add an skip file to skip test for esp32 in ci (funhouse, magtag and metroesp32s2) like this example https://github.com/adafruit/Adafruit_TinyUSB_Arduino/tree/master/examples/MassStorage/msc_sd
> 
> The rest (samd/nrf) seems to be explicit type case which we can fix with the example. If you are still not sure how to do that, don't worry, just add back the ino file. I will take care of the ci (by pushing fix myself)

I’ve reuploaded the ino file. It uses some new contents that I added in the header file (in this PR).
Thank you for hosting this wonderful project and also for helping me to fix this PR!

File: src/class/hid/hid.h [hunk 1]
```diff
@@ -350,6 +350,22 @@ typedef enum
...
 /// @}
 
+//--------------------------------------------------------------------+
+// NKROKeyboard
+//--------------------------------------------------------------------+
+/** \addtogroup ClassDriver_HID_NKROKeyboard NKROKeyboard
+ *  @{ */
+
+/// HID NKROKeyboard Report
+typedef struct TU_ATTR_PACKED
```
Comment (hathach) [inline src/class/hid/hid.h:360]: Could I ask why you decide to choose this layout with 13 key array and 1 custom key ?

File: src/class/hid/hid_device.h [hunk 2]
```diff
@@ -377,6 +377,42 @@ static inline bool  tud_hid_gamepad_report(uint8_t report_id, int8_t x, int8_t y
...
+      HID_INPUT        ( HID_DATA | HID_VARIABLE | HID_ABSOLUTE )  ,\
+      /* 104 Keys as bitmap */ \
+      HID_USAGE_MIN    ( 0                                      )  ,\
+      HID_USAGE_MAX    ( 103                                    )  ,\
+      HID_LOGICAL_MIN  ( 0                                      )  ,\
+      HID_LOGICAL_MAX  ( 1                                      )  ,\
+      HID_REPORT_SIZE  ( 1                                      )  ,\
+      HID_REPORT_COUNT ( 104                                    )  ,\
+      HID_INPUT        ( HID_DATA | HID_VARIABLE | HID_ABSOLUTE )  ,\
+      /* 1 Custom Keyboard Key */ \
```
Comment (hathach) [inline src/class/hid/hid_device.h:406]: why would we have the custom key with max (usage/logical) = 231, could we just merge this with 13 keys, and have them with max (usage/logical) = 255 ( similar to normal key report).

**Review by hathach — COMMENTED** [review]
I have fixed the ci (please pull on your fork before making any changes). I have a few questions though
1. Why would you choose this report layout
2. Why the custom key is special comparing to other 13 keys, could we just have 14 keys with max = 255
3. they keyboard report lack the support for led indicator (caplocks/numlock), should we add it as well ?

File: src/class/hid/hid_device.h (hunk 2 above)
Comment (hlcmOxO) [reply src/class/hid/hid_device.h:406]: This templates is referred from another project: NicoHood/HID. I just transplanted it into this project. And I think you are right about the layout and led issue, I’ll try to improve(or rewrite) the descriptor. And thank you for fixing the ci!

Comment (hlcmOxO) [reply src/class/hid/hid_device.h:406]: the exact path I referred from is https://github.com/NicoHood/HID/blob/master/src/SingleReport/SingleNKROKeyboard.cpp

File: src/class/hid/hid.h (hunk 1 above)
Comment (hlcmOxO) [reply src/class/hid/hid.h:360]: thank you, i've change it to 256key layout (32bytes)

**Comment by hlcmOxO** [comment]
@hathach i have change the layout of the key and add led input. Tested with the new example(already uploaded, keys are working properly) on RP2040.

the problem now is that though i can't get led indicators working (so i didn't code for it in the example).

I try with the code below, the capslock led didn't turn on when the capslock were pressed.

`#include <Arduino.h>
#include <Adafruit_TinyUSB.h>
/*********************************************************************
 Adafruit invests time and resources providing this open source code,
 please support Adafruit and open-source hardware by purchasing
 products from Adafruit!

 MIT license, check LICENSE for more information
 Copyright (c) 2019 Ha Thach for Adafruit Industries
 All text above, and the splash screen below must be included in
 any redistribution
*********************************************************************/


/* This sketch demonstrates USB HID NKRO keyboard.
 * This sketch is only valid on boards which have native USB support
 * and compatibility with Adafruit TinyUSB library.
 * For example SAMD21, RP2040, ATMEGA32U4.
 *
 * Make sure you select the TinyUSB USB stack.
 * You can test the keyboard in a notepad application.
 */

// HID report descriptor using TinyUSB's template
// Single Report (no ID) descriptor
uint8_t const desc_hid_report[] =
    {
        TUD_HID_REPORT_DESC_NKROKEYBOARD()
    };

// USB HID object. For ESP32 these values cannot be changed after this declaration
// desc report, desc len, protocol, interval, use out endpoint
Adafruit_USBD_HID usb_hid(desc_hid_report, sizeof(desc_hid_report), HID_ITF_PROTOCOL_NONE, 2, true);

// Report payload defined in src/class/hid/hid.h
// - 1byte for modifiers
// - 32bytes for keys

hid_nkrokeyboard_report_t kb;

// Output report callback for LED indicator such as Caplocks
void hid_report_callback(uint8_t report_id, hid_report_type_t report_type, uint8_t const* buffer, uint16_t bufsize)
{
  (void) report_id;
  (void) bufsize;

  // LED indicator is output report with only 1 byte length
  if ( report_type != HID_REPORT_TYPE_OUTPUT ) return;

  // The LED bit map is as follows: (also defined by KEYBOARD_LED_* )
  // Kana (4) | Compose (3) | ScrollLock (2) | CapsLock (1) | Numlock (0)
  uint8_t ledIndicator = buffer[0];

  // turn on LED if capslock is set
  digitalWrite(LED_BUILTIN, ledIndicator & KEYBOARD_LED_CAPSLOCK);

#ifdef PIN_NEOPIXEL
  pixels.fill(ledIndicator & KEYBOARD_LED_CAPSLOCK ? 0xff0000 : 0x000000);
  pixels.show();
#endif
}

void setup()
{
#if defined(ARDUINO_ARCH_MBED) && defined(ARDUINO_ARCH_RP2040)
  // Manual begin() is required on core without built-in support for TinyUSB such as mbed rp2040
  TinyUSB_Device_Init(0);
#endif

  Serial.begin(115200);

  // Notes: following commented-out functions has no affect on ESP32
  // usb_hid.setBootProtocol(HID_ITF_PROTOCOL_KEYBOARD);
  // usb_hid.setPollInterval(2);
  // usb_hid.setReportDescriptor(desc_hid_report, sizeof(desc_hid_report));
  // usb_hid.setStringDescriptor("TinyUSB Keyboard");

  usb_hid.setReportCallback(NULL, hid_report_callback);

  usb_hid.begin();

  // led pin
  pinMode(LED_BUILTIN, OUTPUT);
  digitalWrite(LED_BUILTIN, LOW);

  // wait until device mounted
  while (!TinyUSBDevice.mounted())
    delay(1);

  Serial.println("Adafruit TinyUSB HID NKRO keyboard example");
}

void add(uint8_t key_value)
{
  kb.keys[key_value / 8] |= 1 << (key_value % 8);
}

void releaseall()
{
  kb.modifier = 0;
  for (int i = 0; i < 32; i++)
  {
    kb.keys[i] = 0;
  }
}

void send()
{
  usb_hid.sendReport(0, &kb, sizeof(kb));
}

void loop()
{
  if (!usb_hid.ready())
    return;
  
  // Reset buttons
  Serial.println("No pressing buttons");
  releaseall();
  send();
  delay(2000);

  // Press A
  Serial.println("Press A");
  add(HID_KEY_A); // HID_KEY_A, HID_KEY_B...etc defined in src/class/hid/hid.h or see HID Usage Tables
  send();
  delay(2000);
  releaseall();
  send();
  delay(2000);

  // Press B
  Serial.println("Press B");
  add(HID_KEY_B);
  send();
  delay(2000);
  releaseall();
  send();
  delay(2000);

  // Press Enter
  Serial.println("Press Enter");
  add(HID_KEY_ENTER);
  send();
  delay(2000);
  releaseall();
  send();
  delay(2000);

  // Press caps lock
  Serial.println("Press caps lock");
  add(HID_KEY_CAPS_LOCK);
  send();
  delay(2000);
  releaseall();
  send();
  delay(2000);
}
`

other part of the NKROKeyboard is all normal. What's wrong with the code(or maybe some problem with the descriptor? )I will be glad if you can tell me how to get the led indicators working.

**Comment by hlcmOxO** [comment]
i have change the layout of the key and add led input. Tested with the new example(already uploaded, keys are working properly) on RP2040.

the problem now is that though i can't get led indicators working (so i didn't code for it in the example).

I try with the code below, the capslock led didn't turn on when the capslock were pressed.
https://github.com/hlcmOxO/Adafruit_TinyUSB_Arduino/blob/test-led-indicator/examples/HID/NKROKeyboard/NKROKeyboard.ino

other part of the NKROKeyboard is all normal. What's wrong with the code(or maybe some problem with the descriptor? )I will be glad if you can tell me how to get the led indicators working.

**Comment by hlcmOxO** [comment]
@hathach  i have change the layout of the key and add led input. Tested with the new example(already uploaded, keys are working properly) on RP2040.

the problem now is that though i can't get led indicators working (so i didn't code for it in the example).

I try with the code below, the capslock led didn't turn on when the capslock were pressed.
https://github.com/hlcmOxO/Adafruit_TinyUSB_Arduino/blob/test-led-indicator/examples/HID/NKROKeyboard/NKROKeyboard.ino

other part of the NKROKeyboard is all normal. What's wrong with the code(or maybe some problem with the descriptor? )I will be glad if you can tell me how to get the led indicators working.

**Pull request closed without merging**